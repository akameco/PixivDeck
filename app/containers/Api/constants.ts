import { API_REQUEST_SUCCESS_TYPE } from './actionTypes'

export const API_REQUEST_SUCCESS: API_REQUEST_SUCCESS_TYPE =
  'Api/API_REQUEST_SUCCESS'
