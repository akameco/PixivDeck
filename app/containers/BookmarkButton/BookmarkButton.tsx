import * as React from 'react'
import IconButton from 'material-ui/IconButton'
import FavoriteIcon from 'material-ui/svg-icons/action/favorite'

const iconSize = 18
const styles = {
  button: {
    width: 32,
    height: 32,
  },
  icon: {
    width: iconSize,
    height: iconSize,
  },
}
const hoverColor = '#b94343'
export interface Props {
  addBookmark: () => undefined
  deleteBookmark: () => undefined
  isBookmarked: boolean
}

const BookmarkButton = ({
  isBookmarked,
  addBookmark,
  deleteBookmark,
}: Props) => {
  const onClick = isBookmarked ? deleteBookmark : addBookmark
  return (
    <IconButton onClick={onClick} style={styles.button} iconStyle={styles.icon}>
      <FavoriteIcon
        color={isBookmarked ? hoverColor : '#777'}
        hoverColor={hoverColor}
      />
    </IconButton>
  )
}

export default BookmarkButton
