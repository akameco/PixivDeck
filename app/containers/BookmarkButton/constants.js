// @flow
import type {
  ADD_BOOKMARK_REQUEST_TYPE,
  ADD_BOOKMARK_SUCCESS_TYPE,
  ADD_BOOKMARK_FAILER_TYPE,
  DELETE_BOOKMARK_REQUEST_TYPE,
  DELETE_BOOKMARK_SUCCESS_TYPE,
  DELETE_BOOKMARK_FAILER_TYPE,
} from './actionTypes'

export const ADD_BOOKMARK_REQUEST: ADD_BOOKMARK_REQUEST_TYPE =
  'BookmarkButton/ADD_BOOKMARK_REQUEST'
export const ADD_BOOKMARK_SUCCESS: ADD_BOOKMARK_SUCCESS_TYPE =
  'BookmarkButton/ADD_BOOKMARK_SUCCESS'
export const ADD_BOOKMARK_FAILER: ADD_BOOKMARK_FAILER_TYPE =
  'BookmarkButton/ADD_BOOKMARK_FAILER'
export const DELETE_BOOKMARK_REQUEST: DELETE_BOOKMARK_REQUEST_TYPE =
  'BookmarkButton/ELETE_BOOKMARK_REQUEST'
export const DELETE_BOOKMARK_SUCCESS: DELETE_BOOKMARK_SUCCESS_TYPE =
  'BookmarkButton/DELETE_BOOKMARK_SUCCESS'
export const DELETE_BOOKMARK_FAILER: DELETE_BOOKMARK_FAILER_TYPE =
  'BookmarkButton/DELETE_BOOKMARK_FAILER'
