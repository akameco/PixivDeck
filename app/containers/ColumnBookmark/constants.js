// @flow
import type {
  ADD_COLUMN_TYPE,
  ADD_COLUMN_SUCCESS_TYPE,
  SET_NEXT_URL_TYPE,
  FETCH_TYPE,
  FETCH_SUCCESS_TYPE,
  FETCH_FAILRE_TYPE,
  REMOVE_ITEM_TYPE,
  FETCH_NEXT_TYPE,
  FETCH_NEXT_SUCCESS_TYPE,
  FETCH_NEXT_FAILRE_TYPE,
} from './actionTypes'

export const ADD_COLUMN: ADD_COLUMN_TYPE = 'ColumnBookmark/ADD_COLUMN'
export const ADD_COLUMN_SUCCESS: ADD_COLUMN_SUCCESS_TYPE =
  'ColumnBookmark/ADD_COLUMN_SUCCESS'
export const SET_NEXT_URL: SET_NEXT_URL_TYPE = 'ColumnBookmark/SET_NEXT_URL'
export const FETCH: FETCH_TYPE = 'ColumnBookmark/FETCH'
export const FETCH_SUCCESS: FETCH_SUCCESS_TYPE = 'ColumnBookmark/FETCH_SUCCESS'
export const FETCH_FAILRE: FETCH_FAILRE_TYPE = 'ColumnBookmark/FETCH_FAILRE'
export const REMOVE_ITEM: REMOVE_ITEM_TYPE = 'ColumnBookmark/REMOVE_ITEM'
export const FETCH_NEXT: FETCH_NEXT_TYPE = 'ColumnBookmark/FETCH_NEXT'
export const FETCH_NEXT_SUCCESS: FETCH_NEXT_SUCCESS_TYPE =
  'ColumnBookmark/FETCH_NEXT_SUCCESS'
export const FETCH_NEXT_FAILRE: FETCH_NEXT_FAILRE_TYPE =
  'ColumnBookmark/FETCH_NEXT_FAILRE'
