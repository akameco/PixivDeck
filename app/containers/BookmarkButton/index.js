// @flow
import { connect, type Connector } from 'react-redux'
import { createSelector } from 'reselect'
import type { Dispatch } from 'types'
import BookmarkButton, { type Props } from './BookmarkButton'
import { addBookmarkRequest, deleteBookmarkRequest } from './actions'
import { makeIsBookmarked } from './selectors'

const mapStateToProps = createSelector(
  makeIsBookmarked(),
  isBookmarked => ({
    isBookmarked,
  })
)

const mapDispatchToProps = (dispatch: Dispatch, { id }) => ({
  addBookmark() {
    dispatch(addBookmarkRequest(id, 'public'))
  },
  deleteBookmark() {
    dispatch(deleteBookmarkRequest(id))
  },
})

type OP = {
  id: number,
}

const connector: Connector<OP, Props> = connect(
  mapStateToProps,
  mapDispatchToProps
)

export default connector(BookmarkButton)
