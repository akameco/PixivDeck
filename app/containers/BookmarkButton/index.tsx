import { connect, Connector } from 'react-redux'
import { createSelector } from 'reselect'
import { Dispatch } from 'types'
import BookmarkButton, { Props } from './BookmarkButton'
import { addBookmarkRequest, deleteBookmarkRequest } from './actions'
import { makeIsBookmarked } from './selectors'

const mapStateToProps = createSelector(
  makeIsBookmarked(),
  isBookmarked => ({
    isBookmarked,
  })
)

const mapDispatchToProps = (dispatch: Dispatch, { id }) => ({
  addBookmark() {
    dispatch(addBookmarkRequest(id, 'public'))
  },

  deleteBookmark() {
    dispatch(deleteBookmarkRequest(id))
  },
})

interface OP {
  id: number
}
const connector: Connector<OP, Props> = connect(
  mapStateToProps,
  mapDispatchToProps
)
export default connector(BookmarkButton)
