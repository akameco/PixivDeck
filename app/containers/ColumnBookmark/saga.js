// @flow
import type { Saga } from 'redux-saga'
import { put, select, call, takeEvery } from 'redux-saga/effects'
import { addTable } from 'containers/ColumnManager/actions'
import { getMyId } from 'containers/LoginModal/selectors'
import { ADD_BOOKMARK_SUCCESS } from '../BookmarkButton/constants'
import * as column from '../Column/sagas'
import * as Actions from './constants'
import * as actions from './actions'
import type { Action } from './actionTypes'
import type { ColumnId } from './reducer'
import { makeSelectColumn, makeSelectIds } from './selectors'

export function* addColumn({ id }: { id: 'public' }): Saga<void> {
  const ids: Array<?ColumnId> = yield select(makeSelectIds())
  if (ids.every(v => v !== id)) {
    yield put(actions.addColumnSuccess(id))
  }

  yield put(addTable(`bookmark-${id}`, { columnId: id, type: 'BOOKMARK' }))
}

const getEndpoint = (userId, restrict) =>
  `/v1/user/bookmarks/illust?user_id=${userId}&restrict=${restrict}`

export function* fetchBookmark({ id }: { id: 'public' }): Saga<void> {
  const { ids, nextUrl } = yield select(makeSelectColumn(), { id })
  const endpoint = nextUrl ? nextUrl : getEndpoint(yield select(getMyId), id)
  yield call(column.fetchColumn, endpoint, id, { ...actions }, ids)
}

export function* fetchNew({
  restrict: id,
}: {
  restrict: $PropertyType<Action, 'id'>,
}): Saga<void> {
  try {
    const { ids } = yield select(makeSelectColumn(), { id })
    const endpoint = getEndpoint(yield select(getMyId), id)
    yield call(column.fetchNew, { endpoint, id, ids, order: true }, actions)
  } catch (error) {
    yield put(actions.fetchFailre(id, error))
  }
}

export default function* root(): Saga<void> {
  yield takeEvery(Actions.ADD_COLUMN, addColumn)
  yield takeEvery([Actions.FETCH, Actions.FETCH_NEXT], fetchBookmark)
  yield takeEvery(ADD_BOOKMARK_SUCCESS, fetchNew)
}
