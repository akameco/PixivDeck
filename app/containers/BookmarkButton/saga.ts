import { call, put, takeEvery } from 'redux-saga/effects'
import * as api from 'containers/Api/sagas'
import { get } from '../Api/sagas'
import * as columnActions from '../ColumnBookmark/actions'
import * as Actions from './constants'
import * as actions from './actions'
import { Restrict } from './types'

interface Props {
  id: number
  restrict: Restrict
}
export function* bookmark({ id, restrict }: Props) {
  try {
    yield call(api.post, '/v2/illust/bookmark/add', {
      illustId: id,
      restrict,
    })
    yield put(actions.addBookmarkSuccess(id, restrict))
  } catch (error) {
    yield put(actions.addBookmarkFailer(id, error))
  }
}
export function* deleteTask({ id }: Props) {
  try {
    yield call(api.post, '/v1/illust/bookmark/delete', {
      illustId: id,
    })
    yield put(actions.deleteBookmarkSuccess(id))
    yield put(columnActions.removeItem('public', id))
  } catch (error) {
    yield put(actions.deleteBookmarkFailer(id, error))
  }
}

function* success({ id }: Props) {
  try {
    yield call(get, `/v1/illust/detail?illust_id=${id}`, true)
  } catch (error) {}
}

function* root() {
  yield takeEvery(Actions.ADD_BOOKMARK_REQUEST, bookmark)
  yield takeEvery(Actions.DELETE_BOOKMARK_REQUEST, deleteTask)
  yield takeEvery(
    [Actions.ADD_BOOKMARK_SUCCESS, Actions.DELETE_BOOKMARK_SUCCESS],
    success
  )
}

export default root
