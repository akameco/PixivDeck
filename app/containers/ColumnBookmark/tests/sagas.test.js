// @flow
import { takeEvery } from 'redux-saga/effects'
import { ADD_BOOKMARK_SUCCESS } from 'containers/BookmarkButton/constants'
import * as sagas from '../saga'
import * as constants from '../constants'

test('root', () => {
  const gen = sagas.default()
  let next = gen.next()
  expect(next.value).toStrictEqual(
    takeEvery(constants.ADD_COLUMN, sagas.addColumn)
  )

  next = gen.next()
  expect(next.value).toStrictEqual(
    takeEvery([constants.FETCH, constants.FETCH_NEXT], sagas.fetchBookmark)
  )

  next = gen.next()
  expect(next.value).toStrictEqual(
    takeEvery(ADD_BOOKMARK_SUCCESS, sagas.fetchNew)
  )
})

test('fetch next', () => {
  const gen = sagas.fetchBookmark({ id: 'public', type: constants.FETCH })
  const next = gen.next()
  expect(next.value).toMatchSnapshot()
  expect(gen.next({ nextUrl: 'fake' }).value).toMatchSnapshot()
})

test('fetch first', () => {
  const gen = sagas.fetchBookmark({ id: 'public', type: constants.FETCH })
  let next = gen.next()

  expect(next.value).toMatchSnapshot()

  next = gen.next({ id: '1' })
  expect(next.value).toMatchSnapshot()

  next = gen.next({ ids: [1, 2, 3] })
  expect(next.value).toMatchSnapshot()
})

test('new', () => {
  const gen = sagas.fetchNew({ restrict: 'public' })
  let next = gen.next({ id: 'public' })
  expect(next.value).toMatchSnapshot()

  next = gen.next({ ids: [1, 2, 3] })
  expect(next.value).toMatchSnapshot()

  next = gen.next('myid')
  expect(next.value).toMatchSnapshot()
})
