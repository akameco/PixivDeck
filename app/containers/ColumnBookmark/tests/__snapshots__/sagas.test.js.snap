// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch first 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "id": "public",
      },
    ],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`fetch first 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`fetch first 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/v1/user/bookmarks/illust?user_id=[object Object]&restrict=public",
      "public",
      Object {
        "addColumn": [Function],
        "addColumnSuccess": [Function],
        "fetch": [Function],
        "fetchFailre": [Function],
        "fetchNext": [Function],
        "fetchNextFailre": [Function],
        "fetchNextSuccess": [Function],
        "fetchSuccess": [Function],
        "removeItem": [Function],
        "setNextUrl": [Function],
      },
      undefined,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`fetch next 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "id": "public",
      },
    ],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`fetch next 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "fake",
      "public",
      Object {
        "addColumn": [Function],
        "addColumnSuccess": [Function],
        "fetch": [Function],
        "fetchFailre": [Function],
        "fetchNext": [Function],
        "fetchNextFailre": [Function],
        "fetchNextSuccess": [Function],
        "fetchSuccess": [Function],
        "removeItem": [Function],
        "setNextUrl": [Function],
      },
      undefined,
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`new 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "id": "public",
      },
    ],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`new 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`new 3`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      Object {
        "endpoint": "/v1/user/bookmarks/illust?user_id=myid&restrict=public",
        "id": "public",
        "ids": Array [
          1,
          2,
          3,
        ],
        "order": true,
      },
      Object {
        "addColumn": [Function],
        "addColumnSuccess": [Function],
        "fetch": [Function],
        "fetchFailre": [Function],
        "fetchNext": [Function],
        "fetchNextFailre": [Function],
        "fetchNextSuccess": [Function],
        "fetchSuccess": [Function],
        "removeItem": [Function],
        "setNextUrl": [Function],
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
